sentry-graphql-java/src/main/java/io/sentry/graphql/SentryInstrumentation.java

```diff
@@ -0,0 +1,130 @@
+import io.sentry.util.Objects;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public final class SentryInstrumentation extends SimpleInstrumentation {
+  private final @NotNull IHub hub;
+
+  public SentryInstrumentation(final @NotNull IHub hub) {
+    this.hub = Objects.requireNonNull(hub, "hub is required");
+  }
+
+  public SentryInstrumentation() {
+    this(HubAdapter.getInstance());
+  }
+
+  @Override
+  public @NotNull InstrumentationState createState() {
+    return new TracingState();
+  }
+
+  @Override
+  public @NotNull InstrumentationContext<ExecutionResult> beginExecution(
+      final @NotNull InstrumentationExecutionParameters parameters) {
+    TracingState tracingState = parameters.getInstrumentationState();
+    tracingState.setTransaction(hub.getSpan());
+    return super.beginExecution(parameters);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public @NotNull DataFetcher<?> instrumentDataFetcher(
+      final @NotNull DataFetcher<?> dataFetcher,
+      final @NotNull InstrumentationFieldFetchParameters parameters) {
+    // We only care about user code
+    if (parameters.isTrivialDataFetcher()) {
+      return dataFetcher;
+    }
+
+    return environment -> {
+      final TracingState tracingState = parameters.getInstrumentationState();
+      final ISpan transaction = tracingState.getTransaction();
+      if (transaction != null) {
+        final ISpan span = transaction.startChild(findDataFetcherTag(parameters));
+        try {
+          final Object result = dataFetcher.get(environment);
+          if (result instanceof CompletableFuture) {
+            ((CompletableFuture<?>) result)
+                .whenComplete(
+                    (r, ex) -> {
+                      if (ex != null) {
+                        span.setThrowable(ex);
+                        span.finish(SpanStatus.INTERNAL_ERROR);
+                      } else {
+                        span.finish();
+                      }
+                    });
+          } else {
+            span.finish();
+          }
+          return result;
+        } catch (Exception e) {
+          span.setThrowable(e);
+          span.finish(SpanStatus.INTERNAL_ERROR);
+          throw e;
+        }
+      } else {
+        return dataFetcher.get(environment);
+      }
+    };
+  }
+
+  @Override
+  public @NotNull CompletableFuture<ExecutionResult> instrumentExecutionResult(
+      final @NotNull ExecutionResult executionResult,
+      final @NotNull InstrumentationExecutionParameters parameters) {
+    TracingState tracingState = parameters.getInstrumentationState();
+    final ISpan transaction = tracingState.getTransaction();
+
+    final List<GraphQLError> errors = executionResult.getErrors();
+    if (transaction != null && errors != null && !errors.isEmpty()) {
+      transaction.setStatus(SpanStatus.INTERNAL_ERROR);
```

Comment:
that is different than other integrations but might make sense, what do you think @bruno-garcia ?
just wondering if a parent has 10 children, 9 of them were successful and one of them failed, is the parent failed or not?

Reply:
As far as I found out, this method is executed at the end when the execution is over. There can be more than one query sent with single GraphQL request - some of them can fail and be returned in "errors" part of the response. I don't think marking such request that has partially failed as a success reflects the reality.

Perhaps there is already a GraphQL instrumentation in other SDKs and we can just mimic its behavior?

Reply:
the way how I see in all the other SDKs is, if you are able to detect the error of the span that caused the issue, you can indeed mark it as `INTERNAL_ERROR` as we do on the `instrumentDataFetcher` method, but it's parent that comes from `hub.getSpan()` could be even a manually created span by the user that is controlled by them, its up to them to mark as `errored` or `successful, not us, since we `rethrow` the exception, they should do it manually, and if its an automatically created by some other integrations, the exception is gonna bubble up and be marked as `INTERNAL_ERROR` anyway, so I'd avoid messing up with the parent span status.

Reply:
Changing status on parent span removed.